jbpm-flow/src/main/java/org/jbpm/process/instance/context/variable/VariableScopeInstance.java

```diff
@@ -81,76 +87,177 @@ public Object getVariable(String name) {
+                    new Handler<>(
+                            getProcessEventSupport(),
+                            getProcessInstance(),
+                            name,
+                            prefixed(name),
+                            instancePrefixed(name)
+                    ));
+        }
+    }
+
+    protected ProcessEventSupport getProcessEventSupport() {
+        return ((InternalProcessRuntime) getProcessInstance()
+                .getKnowledgeRuntime()
+                .getProcessRuntime())
+                .getProcessEventSupport();
+    }
+
+    private static class Handler<T> implements ReferenceVariableInstance.OnSetHandler<T> {
+
+        transient final ProcessEventSupport processEventSupport;
+        transient final ProcessInstance processInstance;
+        final String name;
+        private final String prefixed;
+        private final String instancePrefixed;
+
+        public Handler(ProcessEventSupport processEventSupport, ProcessInstance processInstance, String name, String prefixed, String instancePrefixed) {
+            this.processEventSupport = processEventSupport;
+            this.processInstance = processInstance;
+            this.name = name;
+            this.prefixed = prefixed;
+            this.instancePrefixed = instancePrefixed;
+        }
+
+        @Override
+        public void before(Object oldValue, Object newValue) {
+            processEventSupport.fireBeforeVariableChanged(
+                    prefixed,
+                    instancePrefixed,
+                    oldValue,
+                    newValue,
+                    processInstance,
+                    processInstance.getKnowledgeRuntime());
         }
-    	if (contextInstanceContainer instanceof CompositeContextNodeInstance) {
-    		this.variableIdPrefix = ((Node) ((CompositeContextNodeInstance) contextInstanceContainer).getNode()).getUniqueId();
-    		this.variableInstanceIdPrefix = ((CompositeContextNodeInstance) contextInstanceContainer).getUniqueId();
-    	}
-	}
 
+        @Override
+        public void after(Object oldValue, Object newValue) {
+            processEventSupport.fireAfterVariableChanged(
+                    prefixed,
+                    instancePrefixed,
+                    oldValue,
+                    newValue,
+                    processInstance,
+                    processInstance.getKnowledgeRuntime());
+        }
+    }
+
+    /**
+     * Assign a value reference to a declared variable instance.
+     * @throws IllegalArgumentException if the given variableName has not been declared in the process
+     */
+    public VariableInstance<?> assignVariableInstance(String variableName, ValueReference reference) {
+        VariableScope variableScope = getVariableScope();
+        Variable variable = variableScope.findVariable(variableName);
+        if (variable == null) {
+            variable = new Variable(variableName);
+            variable.setMetaData("on-the-fly", true);
```

Comment:
What is this on-the-fly used for?

Reply:
so this is currently not used anywhere, but I figured we'd better mark "undeclared" variables somehow (i.e. variables that are defined "on the fly"). This should (of course) be an internal implementation detail. Do we want to 

1) add a new field in Variable ?
2) use some specific meta data field instead of "on-the-fly" ?

Reply:
I'd go for new field so it is more explicit and easier to follow